src/refererDetection.js

```diff
@@ -0,0 +1,140 @@
+import { logError } from './utils';
+
+export function detectReferer(win) {
+  function getLevels() {
+    let levels = walkUpWindows();
+    let ancestors = getAncestorOrigins();
+
+    if (ancestors) {
+      for (let i = 0, l = ancestors.length; i < l; i++) {
+        levels[i].ancestor = ancestors[i];
+      }
+    }
+    return levels;
+  }
+
+  function getAncestorOrigins() {
+    try {
+      if (!win.location.ancestorOrigins) {
+        return;
+      }
+      return win.location.ancestorOrigins;
+    } catch (e) {
+      logError('Error while walking ancestorOrigins', e);
+    }
+  }
+
+  function getPubUrlStack(levels) {
+    let stack = [];
+    let defUrl = null;
+    let encodedUrl = null;
+    let frameLocation = null;
+    let prevFrame = null;
+    let prevRef = null;
+    let ancestor = null;
+    let detectedRefererUrl = null;
+
+    let i;
+    for (i = levels.length - 1; i >= 0; i--) {
+      try {
+        frameLocation = levels[i].location;
+      } catch (e) {
+        logError('Error while accessing cross domain iframe location', e);
+      }
+
+      if (frameLocation) {
+        encodedUrl = encodeURIComponent(frameLocation);
+        stack.push(encodedUrl);
+        if (!detectedRefererUrl) {
+          detectedRefererUrl = encodedUrl;
+        }
+      } else if (i !== 0) {
+        prevFrame = levels[i - 1];
+        try {
+          prevRef = prevFrame.referrer;
+          ancestor = prevFrame.ancestor;
+        } catch (e) {
+          logError('Error while accessing cross domain iframe referrer', e);
+        }
+
+        if (prevRef) {
+          encodedUrl = encodeURIComponent(prevRef);
+          stack.push(encodedUrl);
+          if (!detectedRefererUrl) {
+            detectedRefererUrl = encodedUrl;
+          }
+        } else if (ancestor) {
+          encodedUrl = encodeURIComponent(ancestor);
+          stack.push(encodedUrl);
+          if (!detectedRefererUrl) {
+            detectedRefererUrl = encodedUrl;
+          }
+        } else {
+          stack.push(defUrl);
+        }
+      } else {
+        stack.push(defUrl);
+      }
+    }
+    return {
+      stack,
+      detectedRefererUrl
+    };
+  }
+
+  function walkUpWindows() {
+    let acc = [];
+    let currentWindow;
+    do {
+      try {
+        currentWindow = currentWindow ? currentWindow.parent : win;
+        try {
+          acc.push({
+            referrer: currentWindow.document.referrer || null,
+            location: currentWindow.location.href || null,
+            isTop: (currentWindow == win.top)
+          });
+        } catch (e) {
+          acc.push({
+            referrer: null,
+            location: null,
+            isTop: (currentWindow == win.top)
+          });
+          logError('Error while accessing cross domain iframe.', e);
```

Comment:
Does this need to be a `logError`?  Based on my testing in a cross domain testpage in Chrome, the data still gets collected (due to available information in ancestor) but it also displays this error.  

Should it be a warning instead perhaps?